src/AggregateFunctions/AggregateFunctionSumMap.h

```diff
@@ -80,6 +80,18 @@ class AggregateFunctionSumMapBase : public IAggregateFunctionDataHelper<
 
     void add(AggregateDataPtr place, const IColumn ** columns, const size_t row_num, Arena *) const override
     {
+        // Check if tuple
+        std::unique_ptr<const IColumn *[]> tuple_columns;
+        auto tuple_col = checkAndGetColumn<ColumnTuple>(columns[0]);
+        if (tuple_col)
+        {
+            tuple_columns.reset(new const IColumn*[tuple_col->tupleSize()]);
```

Comment:
We should try not to allocate memory with `new` in these functions, because it's going to be slow. Do I get it right that the tuple is always (key, value) and the size is 2? Maybe we can put tuple_columns on stack instead, and check with `assert` that the size is correct.

Reply:
It is not documented, but sumMap can have unlimited arrays as input. So does the tuple.

Reply:
Oh, I see. The types of ColumnTuple::getColumns and IColumn** are not compatible, so maybe we can make a templated version:
```
template<typename TColumns>
addImpl(...., TColumns columns)
```
and then call it depending on the argument type:
```
void add(...)
{
       if (argument_is_tuple)
       {
               addImpl(..., static_cast<ColumnTuple *>(columns[0])->getColumns());
        }
        else
        {
                addImpl(..., columns);
         }
}

Reply:
Oops, forgot the reference -- must be `addImpl(..., TColumns & columns)`, or we will copy the vector.

Reply:
That's a good idea, done.